src/Analyzer/ArrayJoinNode.cpp

```diff
@@ -0,0 +1,72 @@
+#include <Analyzer/ArrayJoinNode.h>
+
+#include <IO/WriteBuffer.h>
+#include <IO/WriteHelpers.h>
+#include <IO/Operators.h>
+
+#include <Parsers/ASTTablesInSelectQuery.h>
+
+#include <Analyzer/Utils.h>
+
+namespace DB
+{
+
+ArrayJoinNode::ArrayJoinNode(QueryTreeNodePtr table_expression_, QueryTreeNodePtr join_expressions_, bool is_left_)
+    : is_left(is_left_)
+{
+    children.resize(children_size);
+    children[table_expression_child_index] = std::move(table_expression_);
+    children[join_expressions_child_index] = std::move(join_expressions_);
+}
+
+void ArrayJoinNode::dumpTreeImpl(WriteBuffer & buffer, FormatState & format_state, size_t indent) const
+{
+    buffer << std::string(indent, ' ') << "ARRAY_JOIN id: " << format_state.getNodeId(this);
+    buffer << ", is_left: " << is_left;
+
+    buffer << '\n' << std::string(indent + 2, ' ') << "TABLE EXPRESSION\n";
+    getTableExpression()->dumpTreeImpl(buffer, format_state, indent + 4);
+
+    buffer << '\n' << std::string(indent + 2, ' ') << "JOIN EXPRESSIONS\n";
+    getJoinExpressionsNode()->dumpTreeImpl(buffer, format_state, indent + 4);
+}
+
+bool ArrayJoinNode::isEqualImpl(const IQueryTreeNode & rhs) const
+{
+    const auto & rhs_typed = assert_cast<const ArrayJoinNode &>(rhs);
```

Comment:
IQueryTreeNode `isEqual` method compares types of nodes. So here it must be valid to just use `assert_cast` otherwise it is logical error. `isEqualImpl`, `updateTreeHashImpl` should not be public need to update them.